### ISSUE TITLE
hotkeys still keeps running if python process exits

### ISSUE THREAD

**Issue by michaelmesser** [spyoungtech/ahk#257]
`python-ahk-{random}.ahk` continues to run after the python process is killed.
This process consumes 5% CPU.
Is there a way to make the ahk processes stop if the python process is killed?

**Comment by spyoungtech** [comment]
Hmm. Can you explain what you mean exactly by _killed_?

Stopping any running ahk subprocesses is the expected behavior and should happen so long as the Python process exits gracefully. However, if the Python process is _killed_ (like through TASKKILL, end-task with task manager, etc.) or otherwise abruptly stopped, or stopped in a manner that doesn't allow for cleanup (like [`os._exit`](https://docs.python.org/3/library/os.html#os._exit)), there's no opportunity to cleanup the running processes.

For the hotkeys process, this is [done by registering](https://github.com/spyoungtech/ahk/blob/84e8ab353811db17c3392f6f5c4603b4dc60bdff/ahk/_hotkey.py#L338) an [atexit](https://docs.python.org/3/library/atexit.html) handler.

**Comment by michaelmesser** [comment]
issue.py
```py
from ahk import AHK
ahk = AHK()

def caps_lock_down():
    ahk.send_input("{Enter}")

ahk.add_hotkey("CapsLock", caps_lock_down)
ahk.start_hotkeys()
ahk.block_forever()
```

Open terminal
```
 venv/Scripts/python.exe issue.py
```
See the executor AHK process
Press CapsLock
See the random AHK process
Close the terminal window

The `executor.ahk` process remains with 0% CPU
The `python-ahk-{random}.ahk` process remains with 5% CPU

I would expect both to close.
My idea are:
1. to make the AHK processes subprocess of the python process
2. to have the AHK processes check if the python process is still running

**Comment by spyoungtech** [comment]
Thanks, I was able to reproduce the behavior you're seeing. Though, I think the behavior is expected due to the fact that closing the window will abruptly terminate `python.exe` meaning that any cleanup actions, like killing the subprocesses, will not execute.

> to make the AHK processes a subprocess of the python process

I'm not sure I understand this completely. The AHK process is a subprocess of the Python process (for example, as created by `subprocess.Popen`). Unless I'm missing something here.

> to have the AHK processes check if the python process is still running

Yeah. I think this situation could be improved by adding additional handling in the AHK processes. I know it can be solved for the case of the `python-ahk-{random}.ahk` process. I'll have to do some additional digging to see if this can reasonably be handled for the hotkey process in a way that doesn't impact performance.

**Comment by michaelmesser** [comment]
I misunderstood how subprocesses work. I'm not sure what the correct solution is.

**Comment by spyoungtech** [comment]
@michaelmesser this should now be resolved with #258 being merged.

It's presently available as a release candidate that you can install like `pip install ahk==1.5.0rc1`. Let me know if that fixes your problem as expected.
